src/routes/v3/oauth_clients.rs

```diff
@@ -0,0 +1,387 @@
+
+        let mut updated_client = existing_client.clone();
+        let OAuthClientEdit {
+            name,
+            icon_url,
+            max_scopes,
+            redirect_uris,
+        } = client_updates.into_inner();
+        if let Some(name) = name {
+            updated_client.name = name;
+        }
+
+        if let Some(icon_url) = icon_url {
+            updated_client.icon_url = icon_url;
+        }
+
+        if let Some(max_scopes) = max_scopes {
+            updated_client.max_scopes = max_scopes;
+        }
+
+        let mut transaction = pool.begin().await?;
+        updated_client
+            .update_editable_fields(&mut *transaction)
+            .await?;
+
+        if let Some(redirects) = redirect_uris {
+            edit_redirects(redirects, &existing_client, &mut transaction).await?;
+        }
+
+        transaction.commit().await?;
+
+        Ok(HttpResponse::Ok().body(""))
+    } else {
+        Ok(HttpResponse::NotFound().body(""))
+    }
+}
+
+#[get("user/oauth_authorizations")]
+pub async fn get_user_oauth_authorizations(
+    req: HttpRequest,
+    pool: web::Data<PgPool>,
+    redis: web::Data<RedisPool>,
+    session_queue: web::Data<AuthQueue>,
+) -> Result<HttpResponse, ApiError> {
+    let current_user = get_user_from_headers(
+        &req,
+        &**pool,
+        &redis,
+        &session_queue,
+        Some(&[Scopes::USER_OAUTH_AUTHORIZATIONS_READ]),
+    )
+    .await?
+    .1;
+
+    let authorizations =
+        OAuthClientAuthorization::get_all_for_user(current_user.id.into(), &**pool).await?;
+
+    let mapped: Vec<models::oauth_clients::OAuthClientAuthorizationInfo> =
+        authorizations.into_iter().map(|a| a.into()).collect_vec();
+
+    Ok(HttpResponse::Ok().json(mapped))
+}
+
+#[delete("user/oauth_authorizations/{client_id}")]
+pub async fn revoke_oauth_authorization(
+    req: HttpRequest,
+    info: web::Path<String>,
+    pool: web::Data<PgPool>,
+    redis: web::Data<RedisPool>,
+    session_queue: web::Data<AuthQueue>,
+) -> Result<HttpResponse, ApiError> {
+    let current_user = get_user_from_headers(
+        &req,
+        &**pool,
+        &redis,
+        &session_queue,
+        Some(&[Scopes::USER_OAUTH_AUTHORIZATIONS_WRITE]),
+    )
+    .await?
+    .1;
+
+    let client_id = models::ids::OAuthClientId::parse(&info.into_inner())?;
+
+    OAuthClientAuthorization::remove(client_id.into(), current_user.id.into(), &**pool).await?;
+
+    Ok(HttpResponse::Ok().body(""))
+}
+
+fn generate_oauth_client_secret() -> String {
+    ChaCha20Rng::from_entropy()
+        .sample_iter(&Alphanumeric)
+        .take(32)
+        .map(char::from)
+        .collect::<String>()
+}
+
+async fn get_oauth_client_from_str_id(
+    client_id: String,
+    exec: impl sqlx::Executor<'_, Database = sqlx::Postgres>,
+) -> Result<Option<OAuthClient>, ApiError> {
+    let client_id = models::ids::OAuthClientId::parse(&client_id)?.into();
+    Ok(OAuthClient::get(client_id, exec).await?)
+}
```

Comment:
this is something that should be a part of the database func/api. see get string methods in database

Reply:
Oh yeah duh, thanks, will do

Reply:
Actually, on second thought - I don't think that should be the database code's concern. Parsing from strings/json is something we can let actix+serde do automatically using the `Deserialize` implementations we have for the ids. I've reworked things so external ids are automatically deserialized in this way and removed this method.